server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java

```diff
@@ -34,119 +40,324 @@
             }
+            builder.field("allow_no_indices", options.allowEmptyExpressions());
             return builder;
         }
 
-        private static void updateSetForValue(EnumSet<WildcardStates> states, String wildcard) {
-            switch (wildcard) {
-                case "open" -> states.add(OPEN);
-                case "closed" -> states.add(CLOSED);
-                case "hidden" -> states.add(HIDDEN);
-                case "none" -> states.clear();
-                case "all" -> states.addAll(EnumSet.allOf(WildcardStates.class));
-                default -> throw new IllegalArgumentException("No valid expand wildcard value [" + wildcard + "]");
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeBoolean(includeOpen);
+            out.writeBoolean(includeClosed);
+            out.writeBoolean(removeHidden);
+            out.writeBoolean(resolveAlias);
+            out.writeBoolean(allowEmptyExpressions);
+        }
+
+        public static class Builder {
+            private boolean includeOpen = true;
+            private boolean includeClosed = false;
+            private boolean removeHidden = true;
+            private boolean resolveAlias = true;
+            private boolean allowEmptyExpressions = true;
+
+            public Builder() {}
+
+            public Builder(WildcardOptions options) {
+                includeOpen = options.includeOpen;
+                includeClosed = options.includeClosed;
+                removeHidden = options.removeHidden;
+                resolveAlias = options.resolveAlias;
+                allowEmptyExpressions = options.allowEmptyExpressions;
+            }
+
+            /**
+             * Open indices will be included. Defaults to true.
+             */
+            public Builder includeOpen(boolean includeOpen) {
+                this.includeOpen = includeOpen;
+                return this;
+            }
+
+            /**
+             * Closed indices will be included. Default to false.
+             */
+            public Builder includeClosed(boolean includeClosed) {
+                this.includeClosed = includeClosed;
+                return this;
+            }
+
+            /**
+             * Hidden indices will be removed from the result. Defaults to true.
+             */
+            public Builder removeHidden(boolean removeHidden) {
+                this.removeHidden = removeHidden;
+                return this;
+            }
+
+            /**
+             * Aliases will be included in the result. Defaults to true.
+             */
+            public Builder resolveAlias(boolean resolveAlias) {
+                this.resolveAlias = resolveAlias;
+                return this;
+            }
+
+            /**
+             * If true, when each expression does not match any indices we consider it an empty result. If false,
+             * we throw an error. Defaults to true.
```

Comment:
This javadoc implies that this is applied on an expression-by-expression basis; I don't recall off the top of my head if that's true. Was that your intent?

Reply:
Yes, it is on expression by expression basis. For example, assuming there indices in our cluster but non of them has the prefix `missing` `GET *,missing?allow_no_indices=false` throws an `IndexNotFoundException` with message `no such index [missing*`. 

I am going to rephrase it a bit though, I think is a bit confusing.

Reply:
Let me know if it's more clear now.